File: skills/code-architect/references/code-examples/typescript/Command.ts
// Command — encapsulate a request as an object, support undo.

interface Command {
  execute(): void;
  undo(): void;
}

// Receiver holding the state commands act upon.
class Counter {
  value = 0;
}

class IncrementCommand implements Command {
  constructor(private readonly counter: Counter, private readonly by: number) {}
  execute(): void {
    this.counter.value += this.by;
  }
  undo(): void {
    this.counter.value -= this.by;
  }
}

// Invoker records history to support undo.
class CommandInvoker {
  private history: Command[] = [];

  run(command: Command): void {
    command.execute();
    this.history.push(command);
  }

  undoLast(): void {
    const command = this.history.pop();
    command?.undo();
  }
}

function demo(): void {
  const counter = new Counter();
  const invoker = new CommandInvoker();

  invoker.run(new IncrementCommand(counter, 5));
  invoker.run(new IncrementCommand(counter, 3));
  console.log(`value = ${counter.value}`); // 8
  invoker.undoLast();
  console.log(`after undo = ${counter.value}`); // 5
}

demo();

File: skills/code-architect/references/code-examples/swift/Command.swift
// Command — encapsulate a request as an object, support undo.

// Receiver holds the actual state.
final class Counter {
    private(set) var value = 0
    func add(_ amount: Int) { value += amount }
    func subtract(_ amount: Int) { value -= amount }
}

protocol Command {
    func execute()
    func undo()
}

struct AddCommand: Command {
    let counter: Counter
    let amount: Int
    func execute() { counter.add(amount) }
    func undo() { counter.subtract(amount) }
}

// Invoker keeps a history for undo.
final class Invoker {
    private var history: [Command] = []

    func run(_ command: Command) {
        command.execute()
        history.append(command)
    }

    func undoLast() {
        guard let last = history.popLast() else { return }
        last.undo()
    }
}

func runDemo() {
    let counter = Counter()
    let invoker = Invoker()

    invoker.run(AddCommand(counter: counter, amount: 5))
    invoker.run(AddCommand(counter: counter, amount: 3))
    print("after two adds:", counter.value)

    invoker.undoLast()
    print("after undo:", counter.value)
}

runDemo()
